SAK-32151 Sakai Rubrics initial support

ddelblanco opened the pull request:
Rubrics service, without the integration with assignments (see SAK 32152)

Read the readme about information about architecture how to compile and configure.

ern commented on rubrics/api/pom.xml:
```diff
@@ -0,0 +1,38 @@
+    </properties>
+
+    <dependencies>
+        <dependency>
```
lombok dependency is global no need to add.

ern commented on rubrics/api/pom.xml:
```diff
@@ -0,0 +1,38 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-annotations</artifactId>
+            <version>${sakai.jackson.annotations.version}</version>
```
this lib is already provided in shared so you can remove the version here.

ern commented on rubrics/api/src/java/org/sakaiproject/rubrics/model/Criterion.java:
```diff
@@ -0,0 +1,56 @@
+@AllArgsConstructor
+public class Criterion implements Serializable {
+
+	private long id;
```
formatting

ern commented on rubrics/impl/pom.xml:
```diff
@@ -0,0 +1,119 @@
+            <groupId>org.sakaiproject.rubrics</groupId>
+            <artifactId>rubrics-service-client-api</artifactId>
+        </dependency>
+        <dependency>
```
sakai logger is slf4j please update

ern commented on rubrics/impl/pom.xml:
```diff
@@ -0,0 +1,119 @@
+            <artifactId>log4j</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
```
lombok is global no need to include here

ern commented on rubrics/impl/pom.xml:
```diff
@@ -0,0 +1,119 @@
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>${sakai.commons.lang3.version}</version>
```
commons-lang3 already declared in master, remove version

ern commented on rubrics/impl/src/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java:
```diff
@@ -0,0 +1,797 @@
+
+import net.sf.ehcache.Cache;
+
+import org.apache.log4j.Logger;
```
use @Slf4j annotation will add logger

ern commented on rubrics/impl/src/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java:
```diff
@@ -0,0 +1,797 @@
+    private String postRubricResource(String targetUri, String json, String toolId) throws IOException {
+        log.debug(String.format("Post to URI '%s' body:", targetUri, json));
+
+        try {
```
use try with resources here for automatic handling of stream

bradszabo replied:
Hi, Earl. We did have these low level connections identified as clean-up items to refactor to also use Traverson (or RestTemplate directly) for consistency and deal with the resource and connection clean-up. Diego will either do that, or clean this up with a finally block as an interim step.

ern commented on rubrics/impl/src/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java:
```diff
@@ -0,0 +1,797 @@
+    private String putRubricResource(String targetUri,String json, String toolId) throws IOException {
+        log.debug(String.format("PUT to URI '%s' body:", targetUri, json));
+
+        try {
```
try with resources

ern commented on rubrics/impl/src/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java:
```diff
@@ -0,0 +1,797 @@
+    private void deleteRubricAssociation(String query,String toolId) throws IOException {
+
+
+        try{
```
try with resources

ern commented on rubrics/impl/src/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java:
```diff
@@ -0,0 +1,797 @@
+    }
+
+
+    @Getter
```
be nice if these declarations were all moved to the top as that is the pattern used everywhere else.

ern commented on rubrics/pom.xml:
```diff
@@ -0,0 +1,103 @@
+                <version>${sakai.wicket.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.slf4j</groupId>
```
slf4j is a global dep no need to declare

ern commented on rubrics/pom.xml:
```diff
@@ -0,0 +1,103 @@
+                <artifactId>slf4j-api</artifactId>
+                <version>${sakai.slf4j.version}</version>
+            </dependency>
+            <dependency>
```
remove log4j

ern commented on rubrics/pom.xml:
```diff
@@ -0,0 +1,103 @@
+            <dependency>
+                <groupId>commons-lang</groupId>
+                <artifactId>commons-lang</artifactId>
+                <version>${sakai.commons.lang.version}</version>
```
version already declared in master

ern commented on rubrics/pom.xml:
```diff
@@ -0,0 +1,103 @@
+            <dependency>
+                <groupId>commons-configuration</groupId>
+                <artifactId>commons-configuration</artifactId>
+                <version>${sakai.commons.configuration.version}</version>
```
version already declared in master

ern commented on rubrics/services/pom.xml:
```diff
@@ -0,0 +1,193 @@
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <start-class>org.sakaiproject.rubrics.App</start-class>
+        <java.version>1.8</java.version>
+        <lombok.version>1.14.8</lombok.version>
```
remove lombok version and use version in master, feel free to update version in master.

ern commented on rubrics/services/pom.xml:
```diff
@@ -0,0 +1,193 @@
+            <artifactId>h2</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
```
remove lombok as its global

ern commented on rubrics/services/pom.xml:
```diff
@@ -0,0 +1,193 @@
+            <artifactId>json-path</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
```
slf4j is global

ern commented on rubrics/services/pom.xml:
```diff
@@ -0,0 +1,193 @@
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.5</version>
```
version declared in master remove

ern commented on rubrics/services/src/main/java/org/sakaiproject/rubrics/repository/custom/TranslationsController.java:
```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+private class Translations{
```
reformat this

ern commented on rubrics/services/src/main/webapp/sass/sakai-rubrics.scss:
```diff
@@ -0,0 +1,314 @@
+// #sakai-rubrics-manager
```
should be moved into morpheus

bradszabo replied:
Hi, Earl. We will have to discuss this as these styles are part of the [Web Components](https://www.webcomponents.org/), which are [intended to be encapsulated (HTML, CSS and JavaScript as a reusable unit)](https://developers.google.com/web/fundamentals/getting-started/primers/shadowdom). With [Shadow DOM](https://www.w3.org/TR/shadow-dom/) or [Shady DOM](https://www.polymer-project.org/1.0/blog/shadydom) the [components inherit some core theme styles from the container](https://www.webcomponents.org/specs#styling) but are meant provide component specific styles, and to not leak component specific styles to anything else on the page.

There may be parts of the styling that make sense to move into Morpheus, but in general, use of Shadow/Shady DOM with web components is a new paradigm that should slightly change the way styling is handled in Sakai. Central theme aspects will stay in Morpheus, but there should be a place for component specific styles to stay directly with the components, especially given that these components can and should be able to be used and rendered completely outside of Sakai. The Rubrics services as designed can actually integrate and work with any LMS platform that wants to add rubrics functionality. There are a few clean-up items, particularly in the view layer to make this a reality which were deferred given that this is targeted for Sakai, but it is not far off. Diego is scheduling a meeting and I am available tomorrow so we can discuss more at that point.

ern commented on rubrics/services/src/main/webapp/sass/sakai-rubrics-associate.scss:
```diff
@@ -0,0 +1,36 @@
+.rubrics-lightbox {
```
should be moved into the tool in morpheus

ern commented:
This a solid contribution on what Sakai should be aiming for in the future.
My comments are minor in nature.

The only area that may need some discussion is with configuration as it should look to include configuration from sakai which can then be mapped when appropriate.

All in all a very nice contribution...

bradszabo commented:
Thanks, Earl. As I mentioned, Diego or I will get the clean-up items fixed soon and we will connect to discuss a couple of these items within the next couple days. With respect to your overall comment and the configuration mapping, I presume you are mostly referring to having to replicate the database connection and Hibernate configuration with the spring.datasource.* properties. I have adjusted the default rubrics application.properties to use Sakai properties via [Spring Boot configuration property placeholders](http://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html#boot-features-external-config-placeholders-in-properties).
```
# Default to use existing Sakai properties - override if running outside of Sakai
spring.datasource.url=${url@javax.sql.BaseDataSource}
spring.datasource.username=${username@javax.sql.BaseDataSource}
spring.datasource.password=${password@javax.sql.BaseDataSource}
spring.datasource.driver-class-name=${driverClassName@javax.sql.BaseDataSource}
spring.datasource.tomcat.validation-query=${validationQuery@javax.sql.BaseDataSource}
spring.jpa.database-platform=${hibernate.dialect}
spring.jpa.generate-ddl=${auto.ddl}
spring.jpa.show-sql=${hibernate.show_sql}
```
I have also updated the README accordingly. Diego will update this PR with those changes soon.

If there are more complex configuration requirements, we can definitely work on overriding the Spring Boot Environment configuration in the Rubrics [@Configuration](https://docs.spring.io/spring-boot/docs/current/reference/html/howto-properties-and-configuration.html) as necessary but I think the Spring Boot config placeholders should meet our needs to eliminate redundancy when Rubrics is running within the same application server and using the same database as Sakai.

buckett commented on rubrics/tool/src/java/org/sakaiproject/rubrics/tool/MyApplication.java:
```diff
@@ -0,0 +1,165 @@
+
+            // then, return the appropriate IRequestHandler, or "null"
+            // to let another listener handle the exception
+                ex.printStackTrace();
```
Code should never use ex.printStackTrace() but use a logger instead if it's needed.

ddelblanco replied:
Oh yes... we will remove that.

buckett commented on rubrics/services/src/main/resources/application.properties:
```diff
@@ -0,0 +1,189 @@
+
+
+
+rubrics.instance.layouts[flex].language[es-ES].default-rubric.title=Nueva R\u00FAbrica
```
Is having all the properties for different languages going to cause problems for translators who are used to having one properties file per language?

bradszabo replied:
Spring Boot is pretty flexible, these can be separated into discrete property files per language and we can come up with a mechanism to optionally add them to the Spring Boot Environment. It is not using standard ResourceBundle support for these, as they are default data mapped into model types, but I think we can make adjustments to accommodate the common working pattern for translators.

ddelblanco replied:
These is the default values that the institution wants to give to a new rubric that is created by an instructor, so this is configuration. As an institution maybe want to have the default rubric in different languages, we added that option of having different default rubrics based in the languages.  This should not be in the code because, as said, it is institutional configuration.

There are the usual properties files to change the menu labels and other texts of the UI, etc...

bradszabo commented on rubrics/services/gulpfile.js:
```diff
@@ -0,0 +1,147 @@
+var runSequence = require('run-sequence');
+
+// for development, change this to point to your tomcat installation 
+var server_base = "/opt/tomcat";
```
This is the cause of the TravisCI failure. Diego is working on correcting this client code handling to separate the concerns of the web developer's direct use of Gulp during development vs how it is incorporated into the Maven build. It should be shifted to pass in the working target directory either via command line or local configuration (developer) or from the maven-frontend-plugin within a Maven build so that it uses the Maven ephemeral build directory (default target) for all client build tool activities and handle all interim processing such as downloading client dependencies for inclusion, such as the bower_components, and then overlaying the customization to the polymer-sortable.js file. That way everything is processed and packaged into the war artifact. The server_base should only come into play for a web developer who wants to code and refresh the running deployment during rapid development cycles without requiring any server side changes to be built and deployed.

ern commented:
This is going to need some refactoring to use Hibernate 4...

ddelblanco commented:
I've just rebased, modified all to work with the actual trunk solved most of the requirements in the comments. 
I will create JIRAS to:
1. Make the service to depend of the sakai parent pom. Now it depends on the springbootparent, and that's why some dependencies needs to be there. 
2. Make the service to look at sakai properties without any extra configuration. (now it needs to add a parameter in JAVA_OPTS for that)
3. Integrate the webcomponent css with morpheus

buckett commented on rubrics/services/pom.xml:
```diff
@@ -0,0 +1,186 @@
+    <name>Sakai Rubrics - REST API &amp; Web Components</name>
+    <description>Sakai Native Rubrics Service - Provides a REST API and Web Components for Rubrics</description>
+
+    <parent>
```
Having this in here means that people can't customise their Sakai build by editing `master/pom.xml` as they will also need to customise this file as well.

buckett commented:
@ddelblanco Ah missed your comment about making JIRAs, you can import the spring dependencies without pulling in the parent POM:

http://docs.spring.io/spring-boot/docs/current/reference/html/using-boot-build-systems.html#using-boot-maven-without-a-parent

ddelblanco commented:
https://jira.sakaiproject.org/browse/SAK-32331
https://jira.sakaiproject.org/browse/SAK-32333
https://jira.sakaiproject.org/browse/SAK-32332

ern commented on master/pom.xml:
```diff
@@ -75,10 +75,13 @@
     <sakai.commons.lang3.version>3.5</sakai.commons.lang3.version>
     <sakai.commons.fileupload.version>1.3.2</sakai.commons.fileupload.version>
     <sakai.commons-math3.version>3.6.1</sakai.commons-math3.version>
+    <sakai.commons-logging.version>1.2</sakai.commons-logging.version>
```
Remove this as commons-logging is not used anymore and is declared elsewhere.

ddelblanco replied:
https://jira.sakaiproject.org/browse/SAK-32337

ern commented on rubrics/services/pom.xml:
```diff
@@ -0,0 +1,186 @@
+        <rest.assured.version>2.9.0</rest.assured.version>
+        <spring.security.version>4.2.1.RELEASE</spring.security.version>
+        <spring.boot.version>1.5.2.RELEASE</spring.boot.version>
+        <hibernate.version>4.3.11.Final</hibernate.version>
```
This is already declared in master <sakai.hibernate.version>4.3.11.Final</sakai.hibernate.version>

ottenhoff replied:
You can just commit and push an additional commit ..... We will use Github squash and merge

ddelblanco replied:
Ok I will edit that in both branches directly from github. 

The problem is that SAK-32152 depends on this branch, so any change here should be included in that one before the last commit on that one and I'm trying to maintain that. But if it is ok...to add additional commits in the PR, I'll do in that way. Much more easy for me. Thanks

ddelblanco commented:
As I've needed to rebase again... something has been wrong with the PR now. For some reason when I've rebased master all the sakai trunk commits has been added to the PR as if they were part of my changes...   I need to find a way to isolate again just my commits. :(

ddelblanco commented:
Seems that is solved, but I need to be sure. Next Monday I'll check and confirm
